### PR TITLE
11-article-topic-query

### PR DESCRIPTION
**Pull request by Kopfkinos**
this commit will allow /api/articles endpoint to accept a topic query. 
it also will add a util function for generating custom promiseReject objs

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -522,3 +522,34 @@ describe("GET /api/articles (sorting queries)", () => {
       })
   })
 })
+describe("GET /api/articles (topic query)", () => {
+  test("200: responds with arr of only article objs with 'mitch' topic", () => {
```
Comment (nc-jo-watson) [inline __tests__/app.test.js:526]: A better description here would be `responds with an array of article objects with *any valid* topic`

Comment (nc-jo-watson) [reply __tests__/app.test.js:526]: So it's not specific to Mitch, though I'm sure he appreciates it.

File: __tests__/app.test.js [hunk 2]
```diff
@@ -522,3 +522,34 @@ describe("GET /api/articles (sorting queries)", () => {
...
+      .get(`/api/articles?topic=diva`)
+      .expect(404)
+      .then(({ body: { msg } }) => {
+        expect(msg).toBe("Topic doesn't exist! (yet...)")
+      })
+  })
+  test("400: responds with bad request error when passed a topic query that is longer than 200 chars", () => {
+    const tooLongQuery =
+      "tyuuoedgtfmzdgxiedeuhjzgrmnklewwohscakdocqnflrbkuraaudgxjugkgnupxlatogqkejgryqfcrnwhznwrqnybejfqjsxewerpkqliwebtsermzyivujbyjmjtagbbxtbmeqrfohwwlbwmbvdpzewagcawrhthadutffwquwrpqstsmbafkdrgoontyhdpnnovdo"
+
```
Comment (nc-jo-watson) [inline __tests__/app.test.js:547]: This test should be part of your `seed` test, as your table constraints should be restricting the data input.

File: __tests__/app.test.js [hunk 3]
```diff
@@ -522,3 +522,34 @@ describe("GET /api/articles (sorting queries)", () => {
...
+  test("200: responds with arr of only article objs with 'mitch' topic", () => {
+    return request(app)
+      .get(`/api/articles?topic=mitch`)
+      .expect(200)
+      .then(({ body: { articles } }) => {
+        articles.forEach((article) => {
+          expect(article.topic).toBe("mitch")
+        })
+      })
+  })
```
Comment (nc-jo-watson) [inline __tests__/app.test.js:535]: Missing test `200: responds with an empty array when passed a valid topic that has no articles`.

E.g. the topic "paper" in your db, I think 🤔

File: models/articles-models.js [hunk 4]
```diff
@@ -1,30 +1,40 @@
 const db = require("../db/connection")
 const { sort } = require("../db/data/test-data/articles")
+const { generatePromiseReject } = require("./model-utils")
 
 exports.fetchArticleById = (article_id) => {
   return db.query("SELECT * FROM articles WHERE article_id = $1", [article_id]).then(({ rows }) => {
     if (!rows.length) {
-      return Promise.reject({ status: 404, msg: "Path not found" })
+      return generatePromiseReject(404, "article")
```
Comment (nc-jo-watson) [inline models/articles-models.js:8]: This function name tells me nothing about what you are acheiving here.
Perhaps a better name... `generateErrors()` ?

**Review by nc-jo-watson — COMMENTED** [review]
Functionality for this ticket has been tested an implemented well, Connor!

However I question the utility of your util function ... it reduces the readability of your code (from wherever it is used, I wouldn't know what error message is being passed back to the client) and, ultimately, is makes your codebase more verbose (the function itself AND the tests add a lot more lines of code than just having the Promise.rejects directly).
